README
                      ---------------------------------
                      SMALL-FOOTPRINT CIM BROKER (sfcb)
                      ---------------------------------

1. What is sfcb
===============
sfcb is a lightweight CIM daemon (aka CIMOM) that responds to CIM client
requests for system management data and/or performs system management 
tasks. sfcb supports most of the standard CIM XML over http/https protocol.
It is highly modular, allowing functionality to be easily added, removed
or customized for different management applications. sfcb is specifically
targeted for small embedded system that do not have the available CPU,
memory or disk resources to support a full-blown enterprise-level CIMOM.
That said, sfcb runs very well on a regular Linux/Unix system and supports
most of the functions required by CIM clients to manage such the system. 
Supported CIM operations include:

	GetClass
	ModifyClass
	CreateClass
	DeleteClass
	EnumerateClasses
	EnumerateClassNames
	GetInstance
	DeleteInstance
	CreateInstance
	ModifyInstance
	EnumerateInstances
	EnumerateInstanceNames
	InvokeMethod (MethodCall)
	ExecQuery
	Associators
	AssociatorNames
	References
	ReferenceNames
	GetQualifier
	SetQualifer
	DeleteQualifier
	EnumerateQualifiers
	GetProperty
	SetProperty

CIM process and lifecycle indications (aka 'events') are also supported. 

sfcb supports HTTP 'chunking' for enumeration output. This allows large
volumes of data to be send back to the CIM client in 'chunks' as they
are generated, rather than waiting until all the results are available and
then sending back as one huge HTTP response. This feature obviously requires
chunking support by the CIM client.

To minimize runtime resource consumption, sfcb allows providers for multiple
CIM classes to be grouped and run under a single process, as well as the
automatic loading and unloading of idle providers according to 
user-configurable parameters.

sfcb includes a default static instance repository provider that
allows the reading and writing of raw instance data for classes without
their own explicit CIM provider. This code is not yet extensively tested.

  
sfcb is part of the Standards Based Linux Instrumentation for Manageability
(SBLIM) Project hosted by SourceForge. The project home page can be found at:

	http://sblim.sourceforge.net/

Please go the the SBLIM Project home page for more information about sfcb
and/or SBLIM, including design and architecture documents, sample providers,
test files, CIM client programs, etc.

What isn't sfcb
===============
Currently sfcb does not support some of the CIM class operations, 
specifically adding and removing new CIM classes dynamically. Instead,
classes are added/removed/modified offline and require restarting the sfcb
daemon for the changes to take effect. However, there is nothing inherent
in the design of sfcb preventing more sophisticated class module from being
implemented, although probably at the expense of a larger footprint.

sfcb only supports the Open Group standard CMPI provider interface. Several 
CMPI providers covering commonly needed system management resources are 
available from the SBLIM Project.

sfcb does not implement any sophisticated client authorization
support at this time. Instead, this functionality can be added
via writing specical purpose external SBLIM providers. There is a sample
authorization SBLIM provider available from the SBLIM Project home page. 

sfcb does include support for basic authentication (which requires writing
a custom authentication library) and authentication based on client 
certificates which can be combined with basic authentication. See 
21. for details.

 
2. Pre-requisites
=================
sfcb can either be built from a tarball downloaded from the SBLIM download area
or by extracting the sources from the CVS repository.

When building sfcb from a tarball, the only requisite is that curl-devel is installed 
(see below).
In fact even this can be avoided if the package is configured with indication disabled.


Building sfcb from sources extracted from CVS  requires that the following packages 
are already installed on your Linux/Unix system:

	automake >= 1.8
	autoconf >= 2.5.8
	libtool >= 1.5.8
	bison >= 1.85
	curl >= 7.11.1-1
	curl-devel >= 7.11.1-1
        flex

If you want to enable SSL support, the following packages are needed as well:

	openssl > 0.9.7
	openssl-devel > 0.9.7

To check if a package is installed and to determine its version type

	rpm -q <packagename>
		
If you want to enable SLP support, the following package is needed as well:
	openslp (available from http://www.openslp.org)

Make sure all the required packages are installed prior to attempting
to build sfcb or any of the supporting packages. You may need to
install a more recent version than what is on your system.  Missing or
outdated packages will result in a variety of build problems.


3. Quick Install Instructions
=============================
Note: the quick install supports you in setting up an environment including
providers and additional support programs. If you are interested only in 
sfcb refer to the instructions in section 5.

The sfcb package includes an install script to download (via CVS), build and
install sfcb and all its supporting packages. To download the sfcb package
and run the install script:

	cvs -z3 -d:pserver:anonymous:@sblim.cvs.sourceforge.net/cvsroot/sblim co -P sfcb
	cd sfcb
	./installsfcb.sh

You will be prompted for various install options, as well as the root password
when it is needed for system operations that require root access. Most of the
time the default options should work. By default sfcb will download its source
files and supporting packages to $PWD/sfcb/ in the current working directory
and install itself under /usr/local/


4. Change History
=================

See NEWS file of the sfcb package


5. Downloading sfcb and mofc Source
===================================

Option 1:
--------

You can download and extract the tarball containing both sfcb and mofc
ready to build:

	http://sourceforge.net/projects/sblim/

and click on the sblim-sfcb package to download.  Then unpack: 

	tar jxf sblim-sfcb-x.y.z.tar.bz2

You are now ready to build sfcb and mofc (see Section 6 below).

Option 2:
--------

To download the latest (unsupported) version of sfcb from CVS source:

	cvs -z3 -d:pserver:anonymous:@sblim.cvs.sourceforge.net/cvsroot/sblim co -P sfcb

You must also download the required mofc (MOF compiler) package INTO
the sfcb source directory:

	cd sfcb
	cvs -z3 -d:pserver:anonymous:@sblim.cvs.sourceforge.net/cvsroot/sblim co -P mofc

Then run autoconfiscate to setup the autoconf files.  From the sfcb directory:

	./autoconfiscate.sh

You are now ready to build sfcb and mofc (see Section 6 below).


6. Building sfcb and mofc
=========================
sfcb and its supporting packages have been 'autoconfiscated' to simplify
building and installing them on different Linux/Unix systems. In most cases
the autoconf scripts will determine the correct build/compile options for
your system.

After downloading the sfcb and mofc, either from CVS or tarball, build the
packages by:

        ./configure
        make

And as root:

        make install
        make postinstall

NOTE: By default, sfcb is installed with a prefix of /usr/local, which is
acceptable for local installation. If sfcb is installed via RPM, the
files will be in the "standard" locations, so that any occurence of a file
name starting with /usr/local must be replaced according to the rules below:

/usr/local/etc/* -> /etc/*
/usr/local/var/* -> /var/*
/usr/local/* -> /usr/*


7. Trimming sfcb and mofc binary
================================
If you wish to reduce the footprint of sfcb and it supporting libraries to
a bare minimum then run the following commands after building and installing
sfcb and mofc

	strip /usr/local/sbin/sfcbd
	strip /usr/local/lib/libsfc*.so.0.*
	strip /usr/local/bin/sfcbmof

It is recommended that you do this after you have confirmed that sfcb is
installed and working correctly because stripping the binaries will remove
useful debugging information should problems arise. However, stripping the
binaries and libs will significantly reduce their size.

You can use the footprinting utility to evaluate the resource usage of sfcb.

First you must download the ps_mem.py utility from http://www.pixelbeat.org/scripts/ps_mem.py
Place that script anywhere in your path, or in the sfcb CVS directory, and ensure that it is
executable.

Then build and start sfcb as described in sections 6 and 10 

Lastly, run the footprint utility as root:

		make footprint

The resulting report will describe the resource usage. You may then try different build
options and rerun the utility to evaluate their affect. The README.sfcb-shrink file 
gives some guidance and tips for further reducing the footprint.
 
 
8. Build Options
================
There are several build options you can set to customize building sfcb,
such as what optional support to include or remove for a smaller footprint.
These options are set on the ./configure command line; e.g.

	./configure --enable-ssl

Options:
	--disable-debug=[yes|no]
		Disable debug/trace message logging. Default=no
	
	--enable-ssl=[yes|no]
		Enable HTTPS (ie HTTP+SSL) support. Default=no

	--enable-jdbc=[yes|no]
		Enable server-side JDBC support (experimental). Default=no

	--disable-indications=[yes|no]
		Disable indication support. Default=no (ie enable indications)
	
	--enable-slp=[yes|no]
		Enable slp support. Default=no
		
	--disable-qualifierrep=[yes|no]
		Disable support for qualifier operations (Get-,Set-,Delete-,
		EnumerateQualifiers). Default=no

        --disable-default-properties=[yes|no]
                Disable generation of default instance properties for SFCB.
		Default=no

        --enable-optimized-enumeration=[yes|no]
                Enable enumeration call optimization for polymorphic 
		invocations. This option can result in less provider invocations
		but may break providers expecting to be called for leaf classes
		only (and then multiple times). Default=no

        --enable-pam=[yes|no]
                Enable building a basic authentication module based on PAM.
		Default=no

        --enable-ipv6=[yes|no]
                Enable IPv6 address support.  This does not disable connections
		on IPv4 addresses.
		Default=no

        --enable-local-connect-only
                SFCB will only support local connection. Http server will not 
                be built. Default=no

        --enable-tests[=TESTS]
                Enables the test suite with the listed tests. If no tests are
                listed, all tests are enabled.
                Valid tests are "commands,unit,xml,wbemcli,gcov,slp,providers"
                Default=no

        --enable-uds
                Enables UDS authentication for the http server.
                Default=no
                This option may require the GNU_SOURCE compiler flag if
                building under glibc 2.8+. (CFLAGS=-D_GNU_SOURCE ./configure)


9. Downloading the CIM Schema
=============================
When building sfcb and mofc, running the 'make postinstall' step will
download and install the standardized DMTF CIM Schema from www.dmtf.org.
The CIM Schema contains all the class definitions for the CIM object
hierarchy used by sfcb. Without the base CIM Schema sfcb will be unable to
register providers and their implemented classes.

The CIM Schema can be re-installed at any time by running 

        su -c 'make install-cimschema'

or equivalently

        /usr/local/share/sfcb/getSchema.sh -f

This script uses curl to download the schema. Should this script fail for
any reason please go to 'http://www.dmtf.org/standards/cim' and manually
download the latest CIM Schema tarball and extract the schema fileset to
/usr/local/share/sfcb/CIM/.

Again, there are a few possibilities to influence the way the schema is
installed. The following variables can be passed to the configure command:

CIMSCHEMA_SOURCE	URL or path name of CIM Schema Zipfile
CIMSCHEMA_MOF		File name of the top level MOF file
CIMSCHEMA_SUBDIRS	If specified, the ZIP file subdirectory structure
			is preserved.
If CIMSCHEMA_SOURCE points to a local file, curl is not required.
Alternatevily, these environment variables may be set before invoking
getSchema.sh.


10. Starting and Stopping sfcb
==============================
sfcb installs an init.d style script under /usr/local/etc/init.d/ to allow
you to start, stop and restart the sfcb daemon. 

To start up sfcb you must first become root and then type

	/usr/local/etc/init.d/sfcb start

To stop sfcb you must first become root and then type

        /usr/local/etc/init.d/sfcb stop

Alternatively you can manually start the sfcd daemon as root by

	/usr/local/sbin/sfcbd

to run it in the foreground. After starting sfcbd in this way you should see
something like

# /usr/local/sbin/sfcbd
--- sfcbd V0.9.0 started - 31805
--- Using /usr/local/etc/sfcb/sfcb.cfg
--- SSL not configured
--- User authentication disabled
--- Max provider procs: 32
--- initSocketPairs: 80
--- Max Http procs: 8
--- sfcbd HTTP  Daemon V0.9.0 started - 31806 - port 5988
--- Keep-alive timeout disabled
--- ClassProvider for /usr/local/var/lib/sfcb/registration/repository/root/cimv2 (no-version) using 481237 bytes
--- ClassProvider for /usr/local/var/lib/sfcb/registration/repository/root/interop (no-version) using 49507 bytes

with three running processes associated with sfcb. e.g.

	# ps -ef
	...
	root     31805 31392  0 20:54 pts/3    00:00:00 /usr/local/sbin/sfcbd
	root     31806 31805  0 20:54 pts/3    00:00:00 /usr/local/sbin/sfcbd
	root     31807 31805  0 20:54 pts/3    00:00:00 /usr/local/sbin/sfcbd

which correspond to the main  sfcbd process itself (pid 31805), the HTTP codec
(pid 31806) and the class provider (pid 31807). Hit CTRL-C at any time to stop 
sfcb and all its associated processes. In rare cases sfcb might hang and/or not 
shutdown cleanly. Please report this as a bug and include how to reproduce the 
problem. Should this ever happen you can 'kill' sfcbd and all associated 
processes by

	killall -9 sfcbd

This should be done as a last resort - always try a clean shutdown first. 
 
To stop and restart sfcbd type

	/usr/local/etc/init.d/sfcb restart


11. Command Line Options
========================
sfcb has a few commandline options that enable/disable particular runtime
features. They are

	sfcbd [-t N|?] [-d] [-v] [-h] [-k] [-s] [-c FILE] [-l LOGLEVEL] 

where
	-t, --trace-components<N|?>
		Activate component-level tracing messages, where N is a
		OR-ed bitmask integer defining which component to trace.
		Specifying "-t ?" which list all the components and their
		associated integer bitmask. They are

		providerMgr		     1 (0x00001)
		providerDrv		     2 (0x00002)
		cimxmlProc		     4 (0x00004)
		httpDaemon		     8 (0x00008)
		upCalls			    16 (0x00010)
		encCalls		    32 (0x00020)
		ProviderInstMgr		    64 (0x00040)
		providerAssocMgr	   128 (0x00080)
		providers		   256 (0x00100)
		indProvider		   512 (0x00200)
		internalProvider	  1024 (0x00400)
		objectImpl		  2048 (0x00800)
		xmlIn			  4096 (0x01000)
		xmlOut			  8192 (0x02000)
		sockets			 16384 (0x04000)
		memoryMgr		 32768 (0x08000)
		msgQueue		 65536 (0x10000)
		xmlParsing		131072 (0x20000)
		responseTiming		262144 (0x40000)

		A useful one to try that reveals the internal workings of
		sfcbd when processing a CIM request (but also does not
		generate too many messages) is "-tm 2019".

	-d, --daemon	
		Run sfcbd (and child processes) in the background.  

	-v, --version	
		Prints version information and exits.

	-h, --help
		Displays usage infomration and exits.

	-k, --color-trace
		Prints trace output in a different color per process for easier
		debugging.  Not recommended when redirecting output to a file.

	-s, --collect-stats	
		Turn on runtime statistics collecting. If enabled then
		various sfcbd runtime statistics will be written to ./sfcbStat
		Default is no statistics are collected.

	-c, --config-file=<FILE>
		Specify an alternative configuration file (see Section 12).
		Default=/usr/local/etc/sfcb/sfcb.cfg 

    -l, --syslog-level=<LOGLEVEL>
         specify the level for syslog. LOGLEVEL can be LOG_INFO, 
         LOG_DEBUG, or LOG_ERR. LOG_ERR is the default

 

12. Runtime Configuration Options
=================================
After starting up, sfcb will read runtime configuration options from the
file /usr/local/etc/sfcb/sfcb.cfg. For example:

	httpPort:       5988
	enableHttp:     true
	httpProcs:      8
	httpsPort:      5989
	enableHttps:    false
	httpsProcs:     8
	provProcs:      32
    maxMsgLen:      10000000
	doBasicAuth:    false
	basicAuthLib:   sfcBasicAuthentication
	useChunking:    true
    selectTimeout:  5
	keepaliveTimeout: 0
	keepaliveMaxRequest: 10
	sslKeyFilePath: /usr/local/etc/sfcb/file.pem
	sslCertificateFilePath: /usr/local/etc/sfcb/server.pem
    sslCiphers: ALL:!ADH:!LOW:!EXP:!MD5:@STRENGTH
	registrationDir: /usr/local/var/lib/sfcb/registration
	providerDirs: /usr/local/lib/sfcb /usr/local/lib /usr/local/lib/cmpi
    trimWhitespace: true

Config Option Descriptions
--------------------------
httpPort
	Local port that sfcbd should listen to receive HTTP (i.e. insecure)
	requests from CIM clients. Default=5988

enableHttp
	sfcb should/not accept HTTP client connections. Default=true.

httpProcs
	Maximum number of simultaneous HTTP client connections before
	blocking new incoming HTTP requests. Default=8

httpsPort
	Local port that sfcbd should listen to receive HTTPS (i.e. secure)
	requests from CIM clients. Default=5989

enableHttps
	sfcb should/not accept HTTPS client connections. Default=false.

httpLocalOnly
	sfcb should only allow HTTP connections from local interface
        (127.0.0.1 or ::1).  Default=false
	
httpsProcs
	Maximum number of simultaneous HTTPS client connections before
	blocking new incoming HTTPS requests. Default=8

provProcs
	Maximum number of simultaneous provider processes. After this
	point if a new incoming request requires loading a new provider,
	then one of the existing providers will first be automatically
        unloaded. Default=32 

maxMsgLen
    Maximum size in bytes of received messages. Messages larger than
    this size will be rejected.
    Default=10000000

doBasicAuth
	Perform basic authentication on the client userid (obtained from the
	HTTP/HTTPS header) before allowing the request. Default=false
	(i.e. no client authentication performed).

basicAuthLib
	Name of the local library to call to authenticate the client userid.
	Default=/usr/local/lib/libsfcBasicAuthentication.la

useChunking
	Tell sfcbd to use HTTP/HTTPS 'chunking' to return large volumes of
	response data to the client in 'chunks', rather than buffering the 
	data and sending it back all at once.
	
enableSlp
        sfcb should/not advertise wbem services with the DA
        Default=true

selectTimeout
	Specify the maximum time in seconds for the HTTP select to complete.
	Default is 5 seconds.

slpRefreshInterval
	timeinterval in seconds in which the slp process refreshes information

keepaliveTimeout
        Specify the maximum time in seconds an sfcb HTTP process will wait
        between two requests on one connection before terminating. Setting 
        it to 0 will disable HTTP keep-alive. Default=15
	
keepaliveMaxRequest
	Specify the maximum number of consecutive request on one connection.
	Setting it to 0 will effectively disable HTTP keep-alive. Default=10

sslKeyFilePath
	Specify the name of the file containing the private key for the
	server's certificate. The file must be in PEM format and may not
	be passphrase-protected. This file is only required if enableHttps is
	set to true. Default=/usr/local/etc/sfcb/file.pem
	
sslCertificateFilePath
	Specify the name of the file containing the server's certificate. 
	The file must be in PEM format. This file is only required if 
	enableHttps is set to true. Default=/usr/local/etc/sfcb/server.pem

sslCiphers
	Specify the list of SSL ciphers to enable. 
	Default is "ALL:!ADH:!LOW:!EXP:!MD5:@STRENGTH" which disables 
		weak ciphers.

registrationDir
	Specify the registration directory, which contains the provider
	registration data, the staging area and the static repository.
	Default=/usr/local/var/lib/sfcb/registration

providerDirs
	A space separated list of directories where sfcb is looking for provider
	libraries. Default=/usr/local/lib /usr/local/lib/cmpi

providerSampleInterval
	The interval in seconds at which the provider manager is checking for
        idle providers. Default: 30

providerTimeoutInterval
	The interval in seconds before an idle provider gets unloaded by the
	provider manager. Default: 60

providerAutoGroup
	If set to true, all providers residing in the same shared library will
	be executed in the same process unless another group has been specified
  	in the provider registration file. Default: true

slpHostnameLib
    Specify the name of the library to use to determine the SLP hostname. This
    will override the default method. You must implement this library based 
    on the sample provided, sfcSlpHostname.c. You must also specify the 
    --enable-slp-hostname-lib configure option to enable this support.

sslClientCertificate
	Specify the way sfcb handles client certificate based authentication. If set to
	ignore it will not request a certificate from the client. If set to
	accept it will request a certificate from the client but not fail if
	the client doesn't present one. If set to require it will refuse the
	client connection if the client doesn't present a certificate. Default=ignore

certificateAuthLib
	Name of the local library to call for client-certificate based 
	user authentication. Called only if sslClientCertificate is not set 
	to ignore. Default=sfcCertificateAuthentication

traceLevel
	Specify the trace level for sfcb. Can be overriden by setting environment
	variable SFCB_TRACE_LEVEL. Default: 0.

traceMask
	Specify the trace mask for sfcb. Overrides the command line option
	--trace-components. Default: 0.

traceFile
	Specify the trace file for sfcb. Can be overriden by setting environment
	variable SFCB_TRACE_FILE. Default: stderr.

trimWhitespace
    Trim leading and trailing whitespace from XML property values. 
    Whitespace is space, tab, crlf. Any ascii value <= 32.
    Default is true

After making any changes to sfcb.cfg you must restart sfcb or get it to reload
the config file by typing 

        /usr/local/etc/init.d/sfcb reload

or equivalently

	killall sfcbd -HUP


13. Environment Variables
=========================
In addition to the sfcbd command line options and configuration file, several
environment variables can also be set which directly or indirectly affect the
runtime behavior of sfcb. Environment variables affecting sfcb and/or its 
SBLIM providers include:

PATH=/usr/local/bin:/usr/local/sbin
	Path to the sfcbd daemon and utilities.

LD_LIBRARY_PATH=/usr/local/lib/sfcb
	Path to the sfcb runtime libraries. Note: the path to the SBLIM
	provider libraries is specified in the sfcb configuration file and 
	therefore does not need to be on the global LD_LIBRARY_PATH. 
	However, some of the utilities re-use sfcbd libraries, hence this
	needs to be added to LD_LIBRARY_PATH for the utilities to work.
	Alternatively, you can add this path to the system-wide dynamic
	linker/loader config file /etc/ld.so.conf

SFCB_PAUSE_PROVIDER

NEWS
Changes in 1.3.14
=================

New features:

- 3367333 New entry point for authentication library
- 3367332 RHOST Support for PAM
- 3435778 Preserve leading/trailing whitespace in values
- 3433733 CreationClassNames should be supplied
- 3446179 Add a feature to validate CMPI types
- 3484025 Disable indication sequencing
- 3462309 Do not allow multiple instances of sfcb
- 3488688 Don't spawn logger for clients
- 3493493 Add sfcbsem
- 3495060 Verify filter and handler information during subscription
- 3498932 Implement ModifyInstance for CIM_ListenerDestination
- 3367361 Allow for releasing auth handle after CIM request completes
- 3367363 Add CMPIRole Support

Bugs fixed: 

- 3414700 assocClass ignored for assoc upcalls
- 3416055 SIGSEGV sometimes does not print provider name
- 3416167 CIM_InstDeletion not in class repo by default
- 3422650 Associators calls to IndicationService depends on cmpi-base
- 3428806 type=null in authorizeFilter call
- 3419721 reset CMPIValue buffer in str2CMPIValue
- 3478930 Handle SIGPIPE in providerMgr
- 3478933 Sequence Number reset issues
- 3483294 InternalProvider Should Handle SFCB_RegisteredProfile
- 3485438 segfault while retrying to deliver an indication
- 3486814 sfcb crashes while freeing the cimxml response
- 3485830 SFCB Makes Redundant ActivateFilter Calls
- 3487419 TestIndication Provider Does Enable Correctly
- 3487675 Log messages can be concatenated
- 3492362 Problem with whitespace trimming if local connect only
- 3457460 PAM_RHOST set wrong
- 3495343 Bad pointer references in trace statements
- 3416164 Don't write CIM_IndicationService to disk
- 3495801 Correction to 3495343
- 3495789 Makefile improvements for test suite
- 3495806 isChild() can cause a provider mismatch
- 3483200 duplicate indication deliveries
- 3496061 IndicationSubscription May Be Undeletable
- 3484083 Exclude SFCBIndicationID
- 3496383 Faster Return from CBDeliverIndication
- 3416060 sigsegv in tool_mm_flush()
- 3497096 double free during stopProc
- 3497765 Better Error Handling in sfcb Provider Manager and Driver
- 3497950 SFCB Semaphore Handling Improvements
- 3498275 Remove __Namespace
- 3498044 RIEnabled conflicts with settable retries
- 3426027 cimAccountPassThrough does not set return code
- 3498719 Better SSL error reporting for sfcb
- 3471814 segfault in objectImpl.c:131,ClObjectGetClString
- 3499930 Check file repository writes
- 3497209 oom panic: out of memory using fsp indprovider
- 3504607 CMPIRole Support in brokerUpc
- 3506306 Fix count for InvokeMethod requests

Changes in 1.3.13
=================

New features:

- 3406823 Make select timeout a config option
- 3406825 Make ssl cipher list configurable

Bugs fixed: 

- 3386391 HOST_NAME_MAX undefined
- 3393324 sfcb process abort in IndCIMXMLHandlerInvokeMethod
- 3398337 Namespaces of filter and handler for IndSub not checked
- 3400114 sfcb fails to start with ipv6 disabled
- 3412033 New files from 3313833 missed release tarball
- 3413031 fixes to perform check before calling free
- 3414236 updateMethodParamTypes uses incorrect qualifier count
- 3414482 Compiler warning in cimcClientSfcbLocal
- 2727918 blob has leak in EnumQualifiers

Changes in 1.3.12
=================

New features:
- 3190623 Set *IsSettable properties to false for IndicationService
- 3313833 Allow for passthrough for expired user to update password
- 3274893 For SFCB 1.4, set SequenceIdentifier as described in DSP1054 1.2.

Bugs fixed:

- 3261868 Enum CQL filter on NULL cause incorrect rsp
- 3213591 Local intf client abort abnormally on enum filter syntax err
- 3291646 Unused var in __NullEvaluate
- 3202466 Single item on method param of array type rejected via Local
- 3287789 interopServerProvider not conservative when answering reqs
- 3296541 CIM_IndicationSubscription corrupts after delilvery fails
- 3314383 Memory leak in CimResource_Enumerate_EP
- 3047562 [patch] Create bzip2 tarballs on "make dist"
- 3314762 Memory leak in method invoke using SFCB local interface
- 3165732 Add ability for client library to specify configuration file
- 3323422 Unnecessary NULL check in control.c
- 3323427 Unnecessary NULL check in httpAdapter
- 3324214 Remove dead code from providerDrv:loadProvider()
- 3324131 indCIMXMLHandler may return uninitialized CMPIStatus
- 3324168 gatherNameSpacesData() may closedir() a NULL pointer
- 3324192 Erroneous condition in loop in objectImpl:locateParameter()
- 3324198 getObjectPath() may deref a NULL pointer
- 3324247 Unnecessary NULL check in providerDrv
- 3324255 Unnecessary NULL check in __cleanup_mt()
- 3324265 _sfcb_trace may deref NULL pointers
- 3324578 getInstance() up call does not close provider context
- 3323419 sfcb local interface causes zombie/defunct client processes
- 3328763 deactivateFilter() doesn't release CMPI object

Changes in 1.3.11
=================

Bugs fixed:

- 3130727 CMPIPropertyMIFT.setProperty() missing objectpath keys
- 3054618 mofpp is overaggressive when detecting comments
- 3162016 sfcbd Segmentation fault in cleanupCimXmlRequest
- 3172080 deliverIndication does not check for matching FROM classes
- 3185178 Wrong classname for returned instances from execQuery
- 3175770 execQuery returns multiple instances of same object
- 3203290 Basic Auth Should Consider Expired Passwords
- 3202420 CDATA value in a string property is improperly XML escaped
- 3038078 Can create indication filters and destinations w/o key props
- 3206663 Enum CQL filter on NULL cause segfault and incorrect rsp
- 3265502 CMPIContext not released in elementCapabilitiesProvider

Changes in 1.3.10
=================

Bugs fixed:
- 3086143 set cmpi status in getPropQualifier
- 3086145 check datetime format in chars2bin
- 3095884 Indication classes missing SystemName
- 3101155 Failed to load provider libraries due to symbol conflicts
- 3101148 SfcbLocal method parameter type conversion and validation
- 3101154 Daemon clients failed SfcbLocal connect due to permission

trace.c

/*
 * trace.c
 *
 * (C) Copyright IBM Corp. 2005
 *
 * THIS FILE IS PROVIDED UNDER THE TERMS OF THE ECLIPSE PUBLIC LICENSE
 * ("AGREEMENT"). ANY USE, REPRODUCTION OR DISTRIBUTION OF THIS FILE
 * CONSTITUTES RECIPIENTS ACCEPTANCE OF THE AGREEMENT.
 *
 * You can obtain a current copy of the Eclipse Public License from
 * http://www.opensource.org/licenses/eclipse-1.0.php
 *
 * Author:        Adrian Schuur <schuur@de.ibm.com>
 * Based on concepts developed by Heidi Neumann <heidineu@de.ibm.com>
 *
 * Description:
 *
 * Trace support for sfcb.
 *
*/


#include "trace.h"
#include <errno.h>
#include "native.h"
#include <string.h>
#include <time.h>

#include <sys/stat.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <sys/types.h>
#include <unistd.h>
#include <pthread.h>
#include "config.h"

/* ---------------------------------------------------------------------------*/
/*                            trace facility                                  */
/* ---------------------------------------------------------------------------*/

char *processName=NULL;
int providerProcess=0;
int idleThreadId=0;
int terminating=0;

int _sfcb_debug = 0;
unsigned long _sfcb_trace_mask = 0;
char *_SFCB_TRACE_FILE = NULL;

TraceId traceIds[]={ 
  {"providerMgr",       TRACE_PROVIDERMGR},
  {"providerDrv",       TRACE_PROVIDERDRV},
  {"cimxmlProc",        TRACE_CIMXMLPROC},
  {"httpDaemon",        TRACE_HTTPDAEMON},
  {"upCalls",           TRACE_UPCALLS},
  {"encCalls",          TRACE_ENCCALLS},
  {"ProviderInstMgr",   TRACE_PROVIDERINSTMGR},
  {"providerAssocMgr",  TRACE_PROVIDERASSOCMGR},
  {"providers",         TRACE_PROVIDERS},
  {"indProvider",       TRACE_INDPROVIDER},
  {"internalProvider",  TRACE_INTERNALPROVIDER},
  {"objectImpl",        TRACE_OBJECTIMPL},
  {"xmlIn",             TRACE_XMLIN},
  {"xmlOut",            TRACE_XMLOUT},
  {"sockets",           TRACE_SOCKETS},
  {"memoryMgr",         TRACE_MEMORYMGR},
  {"msgQueue",          TRACE_MSGQUEUE},
  {"xmlParsing",        TRACE_XMLPARSING},
  {"responseTiming",    TRACE_RESPONSETIMING},
  {"dbpdaemon",         TRACE_DBPDAEMON},
  {"slp",               TRACE_SLP},
  {NULL,0}
};   

/* 
  set the fg color on the console 
  reset = 1 will restore previous color set
  see trace.h for valid attr, fg, bg values

  TODO: fix reset of fg color.  currently we assume it was white
*/
void changeTextColor(int reset)
{ 
  char command[13];
  int attr = (reset) ? RESET : currentProc % 2;
  int fg = (reset) ? WHITE : currentProc % 7;
  int bg =  BLACK;
  if (fg == BLACK) fg = WHITE;
  sprintf(command, "%c[%d;%d;%dm", 0x1B, attr, fg + 30, bg + 40);
  fprintf(stderr, "%s", command);
}

void _sfcb_trace_start(int n)
{
  if (_sfcb_debug < n) {
    _sfcb_debug = n;
  }
}

void _sfcb_trace_stop()
{
   _sfcb_debug = 0;
}

void _sfcb_trace_init()
{

   char *var = NULL;
   char *err = NULL;
   FILE *ferr = NULL;

   var = getenv("SFCB_TRACE");
   if (var != NULL) {
      _sfcb_debug = atoi(var);
   }
   else {
      _sfcb_debug = 0;
   }

   err = getenv("SFCB_TRACE_FILE");
   if (err != NULL) {
      if ((((ferr = fopen(err, "a")) == NULL) || fclose(ferr))) {
         mlogf(M_ERROR,M_SHOW, "--- Couldn't create trace file\n");
         return;
      }
      _SFCB_TRACE_FILE = strdup(err);
   }
   else {
      if (_SFCB_TRACE_FILE)
         free(_SFCB_TRACE_FILE);
      _SFCB_TRACE_FILE = NULL;
   }
}

char *_sfcb_format_trace(char *fmt, ...)
{
   va_list ap;
   char *msg = (char *) malloc(MAX_MSG_SIZE);
   va_start(ap, fmt);
   vsnprintf(msg, MAX_MSG_SIZE, fmt, ap);
   va_end(ap);
   return msg;
}

void _sfcb_trace(int level, char *file, int line, char *msg)
{

   if (msg == NULL)
     return;

   struct tm cttm;
   struct timeval tv;
   struct timezone tz;
   long sec = 0;
   char *tm = NULL;
   FILE *ferr = NULL;

   if ((_SFCB_TRACE_FILE != NULL)) {
      if ((ferr = fopen(_SFCB_TRACE_FILE, "a")) == NULL) {
         mlogf(M_ERROR,M_SHOW, "--- Couldn't open trace file");
         return;
      }
      colorTrace = 0; /* prevents escape chars from getting into the file */
   }
   else {
      ferr = stderr;
   }

   if (gettimeofday(&tv, &tz) == 0) {
      sec = tv.tv_sec + (tz.tz_minuteswest * -1 * 60);
      tm = (char *) malloc(20 * sizeof(char));
      memset(tm, 0, 20 * sizeof(char));
      if (gmtime_r(&sec, &cttm) != NULL) {
         strftime(tm, 20, "%m/%d/%Y %H:%M:%S", &cttm);
      }

      if (colorTrace) {
        changeTextColor(0);
        fprintf(ferr, "[%i] [%s] %d/%p --- %s(%i) : %s\n", level, tm, currentProc, (void *)pthread_self(), file,
                line, msg);
        changeTextColor(1);
      }
      else {
        fprintf(ferr, "[%i] [%s] %d/%p --- %s(%i) : %s\n", level, tm, currentProc, (void *)pthread_self(), file,
                line, msg);
      }

      free(tm);
      free(msg);
   }

   if ((_SFCB_TRACE_FILE != NULL)) {
      fclose(ferr);
   }

}

extern void _sfcb_set_trace_mask(int n)
{
   _sfcb_trace_mask = n;
}

extern void _sfcb_set_trace_file(char * file)
{
  if (_SFCB_TRACE_FILE) {
    free (_SFCB_TRACE_FILE);
  }
  if (strcmp(file,"stderr") == 0) {
    _SFCB_TRACE_FILE = NULL;
  } else {
    _SFCB_TRACE_FILE = strdup(file);
  }
}

void _sfcb_trap(int tn)
{
#ifdef SFCB_IX86
   char *tp;
   int t;
   if ((tp = getenv("SFCB_TRAP"))) {
      t = atoi(tp);
      if (tn == t)
         asm("int $3");
   }
#endif
}

sigHandler *setSignal(int sn, sigHandler * sh, int flags)
{
   struct sigaction newh, oldh;
   newh.sa_handler = sh;
   sigemptyset(&newh.sa_mask);
   newh.sa_flags = flags;

   if (sn == SIGALRM)
      newh.sa_flags |= SA_INTERRUPT;

   if (sigaction(sn, &newh, &oldh) < 0)
      return SIG_ERR;

   return oldh.sa_handler;
}

#ifdef UNITTEST
// Embedded unittest routine
int trace_test()
{
    int fail=0;
    _sfcb_trace_start(1);
    if (_sfcb_debug < 1 ) {
        printf("  _sfcb_trace_start() test failed.\n");
        fail=1;
    }
    _sfcb_trace_stop();
    if (_sfcb_debug != 0  ) {
        printf("  _sfcb_trace_stop() test failed.\n");
        fail=1;
    }
    _sfcb_trace_init();
    if (_SFCB_TRACE_FILE == NULL  ) {
        printf("  _sfcb_trace_init() test failed.\n");
        fail=1;
    }
    int num=99;
    char *msg=_sfcb_format_trace("%s %d","Test message",num);
    if (strcmp(msg,"Test message 99")) {
        printf("  _sfcb_format_trace() test failed, message=%s.\n",msg);
        fail=1;
    }
    
    
    return fail;
}
#endif

